File: core/repository/semantic/semantic_service_engine.py
from __future__ import annotations

from typing import Dict, List


def infer_semantic_services(topology: Dict[str, List[str]], imports: List[str], routes: List[str], dependencies: List[str]) -> Dict[str, List[str]]:
    modules = sorted(set((topology or {}).get("modules", [])))
    services = sorted({m for m in modules if "/" in m or "\\" in m})
    apis = sorted(set(r.split()[0] if " " in r else r for r in routes or []))
    queues = sorted([d for d in (dependencies or []) if d.lower() in {"celery", "rq", "kafka", "rabbitmq"}])
    gateways = sorted([imp for imp in (imports or []) if imp.lower().endswith(("gateway", "client"))])
    databases = sorted([d for d in (dependencies or []) if d.lower() in {"sqlalchemy", "psycopg2", "mongoose", "typeorm"}])
    return {
        "services": services,
        "apis": apis,
        "workers": sorted([s for s in services if s.lower().endswith(("worker", "job"))]),
        "schedulers": sorted([s for s in services if s.lower().endswith(("scheduler", "cron"))]),
        "queues": queues,
        "gateways": gateways,
        "databases": databases,
    }

File: src/repository/semantic/semanticServiceEngine.ts
/**
 * Converted from Python: core/repository/semantic/semantic_service_engine.py
 * @generated — WebWeaveX python→javascript library port
 */

import * as py from "../../runtime/pyCompat.js";

export function inferSemanticServices(topology: any, imports: any, routes: any, dependencies: any): any {
  var modules: any = py.sorted(py.toSet(py.get(py.or2(topology, () => ({})), "modules", [])));
  var services: any = py.sorted(py.toSet(py.iter(modules).filter((m: any) => (py.contains(m, "/") || py.contains(m, "\\"))).map((m: any) => m)));
  var apis: any = py.sorted(py.toSet(py.iter(py.or2(routes, () => ([]))).map((r: any) => (py.contains(r, " ") ? py.at(py.split(r), 0) : r))));
  var queues: any = py.sorted(py.iter(py.or2(dependencies, () => ([]))).filter((d: any) => py.contains(new Set(["celery", "rq", "kafka", "rabbitmq"]), String(d).toLowerCase())).map((d: any) => d));
  var gateways: any = py.sorted(py.iter(py.or2(imports, () => ([]))).filter((imp: any) => py.truthy(py.endswith(String(imp).toLowerCase(), ["gateway", "client"]))).map((imp: any) => imp));
  var databases: any = py.sorted(py.iter(py.or2(dependencies, () => ([]))).filter((d: any) => py.contains(new Set(["sqlalchemy", "psycopg2", "mongoose", "typeorm"]), String(d).toLowerCase())).map((d: any) => d));
  return {"services": services, "apis": apis, "workers": py.sorted(py.iter(services).filter((s: any) => py.truthy(py.endswith(String(s).toLowerCase(), ["worker", "job"]))).map((s: any) => s)), "schedulers": py.sorted(py.iter(services).filter((s: any) => py.truthy(py.endswith(String(s).toLowerCase(), ["scheduler", "cron"]))).map((s: any) => s)), "queues": queues, "gateways": gateways, "databases": databases};
}
